share/github-backup-utils/ghe-rsync

```diff
@@ -13,15 +13,30 @@ set -o pipefail
 # Filter vanished file warnings from both stdout (rsync versions < 3.x) and
 # stderr (rsync versions >= 3.x). The complex redirections are necessary to
 # filter stderr while also keeping stdout and stderr separated.
-IGNOREOUT='^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
-(rsync "${@}" $GHE_EXTRA_RSYNC_OPTS 3>&1 1>&2 2>&3 3>&- |
-  (egrep -v "$IGNOREOUT" || true)) 3>&1 1>&2 2>&3 3>&- |
-  (egrep -v "$IGNOREOUT" || true)
+ignoreout='^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
+
+# Remove --ignore-missing-args parameter if using an earlier version of rsync.
+if ! rsync -h | grep -q '\-\-ignore-missing-args'; then
+  for parameter; do
+    [[ ! $parameter == "--ignore-missing-args" ]] && parameters+=("$parameter") || ignore23=1
```

Comment:
Nit: https://github.com/github/backup-utils/blob/master/STYLEGUIDE.md#scripts-should-not-use-bash-arrays

Reply:
Still trying to find an alternative to this.

So far on all my attempts, if I try and convert the arguments to a string so to then remove `--ignore-missing-args`, it is stripping the quotes out, so things like `--rsync-path='sudo -u git rsync'` become `--rsync-path=sudo -u git rsync`.

Reply:
For the life of me, I can't find where I've dealt with this situation in the past and can't think of a solution without resorting to array manipulation or getting into hacking brittle and confusing POSIX compliant solutions.

I think we can make an exception in this situation as we only want to remove an element from the `$@` array.

That said, I'm not sure this is doing as it should be when run on hosts _with_ support for this option:

```
$ rsync -h | grep '\-\-ignore-missing-args'
     --ignore-missing-args   ignore missing source args without error
$
$ bash -x ./share/github-backup-utils/ghe-rsync --rsync-path=\'sudo -u git rsync\' --ignore-missing-args
[...]
+ ignoreout='^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
+ rsync -h
+ grep -q '\-\-ignore-missing-args'
+ for parameter in '"$@"'   <---- **I DONT THINK WE SHOULD BE HERE**
+ [[ ! --rsync-path='sudo == \-\-\i\g\n\o\r\e\-\m\i\s\s\i\n\g\-\a\r\g\s ]]
+ parameters+=("$parameter")
+ for parameter in '"$@"'
+ [[ ! -u == \-\-\i\g\n\o\r\e\-\m\i\s\s\i\n\g\-\a\r\g\s ]]
+ parameters+=("$parameter")
+ for parameter in '"$@"'
+ [[ ! git == \-\-\i\g\n\o\r\e\-\m\i\s\s\i\n\g\-\a\r\g\s ]]
+ parameters+=("$parameter")
+ for parameter in '"$@"'
+ [[ ! rsync' == \-\-\i\g\n\o\r\e\-\m\i\s\s\i\n\g\-\a\r\g\s ]]
+ parameters+=("$parameter")
+ for parameter in '"$@"'
+ [[ ! --ignore-missing-args == \-\-\i\g\n\o\r\e\-\m\i\s\s\i\n\g\-\a\r\g\s ]]
+ ignore23=1
+ egrep -v '^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
+ rsync '--rsync-path='\''sudo' -u git 'rsync'\'''  <---- **THE ARG HAS GONE AWOL**
+ egrep -v '^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
rsync: link_stat "/Users/lildude/github/backup-utils/git" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1209) [sender=3.1.3]
+ res=23
+ '[' 23 = 24 ']'
+ '[' 23 = 23 ']'
+ '[' -n 1 ']'
+ res=0
+ exit 0
$
```

Reply:
> That said, I'm not sure this is doing as it should be when run on hosts with support for this option:

I'm struggling to reproduce this, but will keep trying.

```
$ rsync -h | grep '\-\-ignore-missing-args'
     --ignore-missing-args   ignore missing source args without error

$ bash -x ./share/github-backup-utils/ghe-rsync --rsync-path=\'sudo -u git rsync\' --ignore-missing-args
[...]
+ ignoreout='^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
+ rsync -h
+ grep -q '\-\-ignore-missing-args'
+ parameters=("$@")
+ egrep -v '^(file has vanished: |rsync warning: some files vanished before they could be transferred)'
+ rsync '--rsync-path='\''sudo' -u git 'rsync'\''' --ignore-missing-args
[...]
```